Fail when statement produces error unexpectedly

I decided to dogfood DocTest when writing the README for [StringLocationConverter](https://github.com/SwiftDocOrg/StringLocationConverter): 

https://github.com/SwiftDocOrg/StringLocationConverter/blob/6cf6579c191e20a5a77a7e3176d37a8d654c9fc4/README.md#usage

The first time I ran `swift doctest --package README.md`, I got the same results as @schwa in #19:

```
TAP version 13
1..2
not ok 1 - `position.line` did not produce `6`
  ---
  column: 69
  file: README.md
  line: 18
  ...
  
not ok 2 - `position.column` did not produce `23`
  ---
  column: 15
  file: README.md
  line: 19
  ...
```

It was only after launching the REPL and copy-pasting the example code that I discovered the underlying issue — I was missing an import statement for the Foundation framework:

```
error: repl.swift:15:20: error: value of type 'String' has no member 'range'
let range = string.range(of: "reason")!
```

Instead of returning `not ok` when an assertion fails to produce a value, the test runner should first check for any output for standard error, and fail if any errors occurred.

--- a/Sources/DocTest/Statement.swift
+++ b/Sources/DocTest/Statement.swift
@@ -16,17 +16,42 @@ public class Statement {
     }
 
     public func tests(with result: Result<String, REPL.Error>) -> [Test] {
-        let metadata: [String: Any] = [
+        var metadata: [String: Any] = [
             "file": self.sourceLocation.file as Any?,
             "line": self.sourceLocation.line as Any?,
             "column": self.sourceLocation.column as Any?
         ].compactMapValues { $0 }
 
-        return expectations.map { expectation in
-            switch expectation {
-            case .value(let expected):
+        switch result {
+        case .failure(let error):
+            metadata["actual"] = error.description
 
-                if case .success(let actual) = result {
+            if expectations.isEmpty {
+                return [test { .failure("- `\(self.code)` produced an error", directive: nil, metadata: metadata) }]
+            } else {
+                return expectations.map { expectation in
+                    switch expectation {
+                    case .value(let expected):
+                        metadata["expected"] = expected
+
+                        return test {
+                            .failure("- `\(self.code)` produced an error", directive: nil, metadata: metadata)
+                        }
+                    case .error:
+                        return test {
+                            .success("- `\(self.code)` produced an error, as expected", directive: nil, metadata: metadata)
+                        }
+                    }
+                }
+            }
+        case .success(let actual):
+            metadata["actual"] = actual
+
+            return expectations.map { expectation in
+                switch expectation {
+                case .value(let expected):
+                    metadata["expected"] = expected
+
                     if actual == expected {
                         return test {
                             .success("- `\(self.code)` produces `\(actual)`", directive: nil, metadata: metadata)
@@ -36,18 +61,7 @@ public class Statement {
                             .failure("- `\(self.code)` produces `\(actual)`, expected `\(expected)`", directive: nil, metadata: metadata)
                         }
                     }
-
-                } else {
-                    return test {
-                        .failure("- `\(self.code)` did not produce `\(expected)`", directive: nil, metadata: metadata)
-                    }
-                }
-            case .error:
-                if case .failure = result {
-                    return test {
-                        .success("- `\(self.code)` produced an error, as expected", directive: nil, metadata: metadata)
-                    }
-                } else {
+                case .error:
                     return test {
                         .failure("- `\(self.code)` didn't produce an error, which was unexpected", directive: nil, metadata: metadata)
                     }

--- a/Tests/DocTestTests/DocTestTests.swift
+++ b/Tests/DocTestTests/DocTestTests.swift
@@ -7,6 +7,8 @@ final class DocTestTests: XCTestCase {
         1 + 1 // => Int = 2
         1 + 1 // => String = "wat"
         1 / 0 // !! Error
+        invalid
+        1 + 1 // => Int = 2
         """#
 
         let expectation = XCTestExpectation()
@@ -17,10 +19,13 @@ final class DocTestTests: XCTestCase {
             case .failure(let error):
                 XCTFail("\(error)")
             case .success(let report):
-                XCTAssertEqual(report.results.count, 3)
+                XCTAssertEqual(report.results.count, 5)
                 XCTAssertTrue(try! report.results[0].get().ok) // 1 + 1 => 2
                 XCTAssertFalse(try! report.results[1].get().ok) // 1 + 1 => "wat"
                 XCTAssertTrue(try! report.results[2].get().ok) // 1 / 0 !! Error
+                XCTAssertFalse(try! report.results[3].get().ok) // invalid
+                XCTAssertTrue(try! report.results[4].get().ok) // 1 + 1 => 2
+
                 expectation.fulfill()
             }
         }

